### ISSUE TITLE
Document lifetimes of Client etc.

### ISSUE THREAD

**Issue by ThiefMaster** [OpenIDC/pyoidc#673]
For example, the Client stores information in `self.state2nonce` and uses that later to validate that the nonce did not change - and if it's not present, that check is silently skipped. Does this have security implications? Honestly, I don't know.

The [client example](https://pyoidc.readthedocs.io/en/latest/examples/rp.html) in the docs does store state and nonce in the session, but never reads back the nonce...

Except for very small applications where you only have a single process/thread, chances are good that a webapp using this library runs on multiple servers or simply with multiple processes - so different Client instances will generate the request and process it, resulting in no match in `state2nonce`. Does this result in a security risk? Again, I don't know, and I think when just using the library I shouldn't *need* to know it (the whole point of a library like this is to not have to know all the details of the protocol).

---

It would be nice if there was clear documentation on what kind of persistent internal state the Client stores, and whether it's safe to reuse a Client or if it's better to create a new Client whenever you want to authenticate a user.

Also, I think it would be nice if there was some documentation on what a user of the library needs to validate or simply be explicit, e.g. requiring you to pass the original nonce to the function that validates it instead of relying on something stored within the Client. That way you are forced to keep track of it (e.g. using a session) and passing it back, instead of silently skipping a check that might be important.

**Comment by tpazderka** [comment]
Yes, unfortunately the documentation is mostly lacking behind the implementation... :(

From my experience, it is safe to reuse the Client instances for different users. The internal states are separated by Grant objects and there shouldn't be any leakage or mismatch.

The issue with `state2nonce` should be fixed on current master if I am not mistaken...

**Comment by JEphron** [comment]
This is pretty important. How can we use this library in a distributed environment if the `Client` keeps state?

**Comment by tpazderka** [comment]
You should be able to use `state` as your key for storing configuration. Returning `state` is mandatory if it was included in request and that should enable your fresh `Client` instance to do `Client.restore()`.

The other option should be to point `SessionDatabase` to some common storage and the code should be able to handle that.

If not, then please file a bug with that.

**Comment by JEphron** [comment]
I don't see a `Client.restore()`, but I do see `Consumer.restore()`. I assume that's what you meant. 
This is the first time I've heard of `Consumer`. It's not in the docs at all, so I had no idea it existed. Even looking at the source now, it's not clear how it's meant to be used. Are there any examples of an RP implemented using `Consumer`?

**Comment by tpazderka** [comment]
Yes, `Consumer` basically provides higher level functions (`begin`, `complete` and `get_user_info`) which should be able to handle most of the regular flow.

And yes, we should rewrite the documentation using the `Consumer` as it it should be easier to implement.

You can get some basic idea from the corresponding tests: https://github.com/OpenIDC/pyoidc/blob/213f20e045faf0fc3902572f62cf784aca674ff4/tests/test_oic_consumer.py
